### ISSUE TITLE
armhf Packages from the Icinga build system

### ISSUE THREAD

**Issue by webdandy** [Icinga/icinga-packaging#20]
Just wanted to know if you have removed the armhf Packages for icinga-jessie or if this is a temp problem?

W: Fehlschlag beim Holen von http://packages.icinga.org/debian/dists/icinga-jessie/InRelease  Erwarteter Eintrag »main/binary-armhf/Packages« konnte in Release-Datei nicht gefunden werden (falscher Eintrag in sources.list oder missgebildete Datei).

Thanks and best regards,

**Comment by mrsmn** [comment]
@webdandy Looks like ```armhf``` packages for ```jessie``` were indeed removed: http://packages.icinga.com/debian/dists/icinga-jessie/main/

I've asked about ```armhf``` packages for ```stretch``` in #15 but have not gotten a response yet.

**Comment by webdandy** [comment]
@mrsmn  Yeah, this is what I also figured out. There are just only packages for i386 and amd64.
But with my icinga source.list I was able to install tha armhf packages last week...

**Comment by mrsmn** [comment]
Knowing the icinga guys, the probably have a good reason for removing it 😄. Let's wait and see.

**Comment by lazyfrosch** [comment]
Hey guys,
unfortunatly yes.

We relied on debmon.org in the past, to maintain, ship and did mirror their packages.

As we are moving to our own build system for `*.deb` packages as well. I migrated the repositories yesterday. Icinga 2.7.0 will be the first release shipped directly.

We haven't decided yet, if or how to provide armhf builds. Mainly because we are not sure, how big the user base is.

You currently have the option to directly use debmon.org for your armhf devices.

If you are a user of armhf and want use to support it, please subscribe to this post and add a thumbs up reaction (the smiley icon above this comment).

Also see: https://www.icinga.com/2017/05/29/news-on-the-icinga-build-system/

**Comment by sysadmama** [comment]
+1 (for every Raspi I own!)

**Comment by lazyfrosch** [comment]
**Please** avoid `+1` comments, they don't help no one, but spam all subscribers:

Click ![bildschirmfoto von 2017-06-22 10-22-07](https://user-images.githubusercontent.com/121874/27424243-ae0ab6b2-5734-11e7-8152-015e56b6c024.png) on my comment

and subscribe to this post: 
![bildschirmfoto von 2017-06-22 10-23-04](https://user-images.githubusercontent.com/121874/27424286-d36b95a2-5734-11e7-853e-aa4a1c41bd4d.png)

**Comment by lesinigo** [comment]
Thumbupped and subscribed.

You guys probably have a better view on your real user base than what I can guess, but imho there will be more and more requests to run Icinga2 on arm as time goes by, mostly for the agent.

But, we also have a couple of RasPi3 (or similar boards) systems happily running a complete monitoring solution with Icinga2, InfluxDB and Grafana and I must say that it's a nice setup when you need some cheap and small system to put on-site somewhere and you don't need the reliability of a "true" server (even if in something like 8 months we never had a single failure).

**Comment by centertivevier** [comment]
It ist a good tool for my raspberry

**Comment by PLanB2008** [comment]
I'd really like to see you guys keep on building packages for raspberry pi. Is there a reason like huge amounts of work to build in armhf support in your new building system you stopped supporting it?

**Comment by dnsmichi** [comment]
Packages for armhf were never built by the former and the new build system; just mirrored from debmon. It takes time to buy hardware, and ensure that builds a properly run. I'm pretty sure @lazyfrosch will take care.

**Comment by phlipse** [comment]
Are there any news on that topic?

**Comment by mvillarejo** [comment]
it would be great to have arm builds back... this is perfect for my raspberry pi

**Comment by jbein** [comment]
Jep! Would be great to have them. Any news regarding ARM builds yet?

**Comment by Nunak** [comment]
We wanted!  Any news regarding ARM builds yet?

**Comment by mzac** [comment]
Also looks like debmon.org packages haven't been updated since 2.7.0-1... Not sure when 2.8.0 packages will be avail?  I also run Icinga2 on a few Raspberry PIs

**Comment by mzac** [comment]
I'm trying to figure out how I could create the debian package for ARM-Hf, it must not be too complicated?  I have never done that before... @dnsmichi in terms of hardware needed to create the ARM builds, would a raspberry pi be all that would be needed?

**Comment by phlipse** [comment]
If somebody has a working build script, I could build for ARMv7. I have an no anymore used board lying around.

**Comment by lazyfrosch** [comment]
Well anyone should be able to build the package, Debian sources are here: https://github.com/Icinga/icinga-packaging/tree/deb/release/icinga2

Problem is full integration in our build and release process. We think about how and when to include that. But nothing has been decided yet.

You will hear it here first.

**Comment by mzac** [comment]
I found this and will git it a try for my Raspberry PIs

https://gist.github.com/andrenarchy/80d9b2e951dfaea7b6f4

**Comment by Nunak** [comment]
Any progress about repository or possibility for debian stretch and arm (RPIs)?

**Comment by simonszu** [comment]
I am using this issue as a reason to install and play around with Jenkins. So far, i have made a successful build. I am planning to create a .deb File for Debian Stretch. If my Raspbian can successfully upgrade Icinga2 from 2.7 to 2.8 with this file, i will link it here. I do not have the resources or the time to create a "real" repository, but at least it will be an intermediate solution to get 2.8 on our raspis.

**Comment by mbuiro** [comment]
Any update on this? Seems like Pi users have been forgotten..

**Comment by Toriniasty** [comment]
Hi chaps,
I've compiled them some time ago (2.8.0-1 version).
If you want I've put them on my onedrive:
https://1drv.ms/f/s!Aq3gq6mIQqpshYZJThqMzSiP7ItupQ

Link will expire in April as probably till that time there will be something newer..

**Comment by Nunak** [comment]
Toriniasty: Many thanks great work! I can confirm it works also on debian :) BTW did you have compiled also icingaweb2 ?

**Comment by Toriniasty** [comment]
Oh yeah, I forgot to mention they were compiled on raspbian.
Regarding icingaweb2 the answer is no. I'm running icingaweb2 on seperate server so I'm not that bothered.

**Comment by simonszu** [comment]
The gist @mzac posted doesn't work for icingaweb2 on Debian Stretch, though. Stretch uses php7.0 by default, and the icingaweb build package tries to build against php5

**Comment by mbuiro** [comment]
Torinasty: Really appreciate the great work, thank you!

**Comment by GringoDeLobo** [comment]
No news? I too would appreciate to be able to install Icinga2 on a raspberry pi via the repository.

**Comment by Codelica** [comment]
Please do consider bringing back ARM builds.   IOT devices like Pi's make ideal remote Icinga nodes for people like us, who have to monitor various customer installations on-site.  As an arch, it's use will only increase!  👍

**Comment by ftrojahn** [comment]
For others looking after information about debmon.org:

https://twitter.com/icinga/status/990167971879358464
https://twitter.com/formorer/status/989475559582420995

**Comment by nbuchwitz** [comment]
I created a build pipeline for Debian ``jessie`` and ``stretch`` packages of Icinga 2 and Icingaweb 2. The packages can be found here:

https://packages.enda.eu/

Please note that these are not official packages, although they are built according to the same specs. I'll do my best to keep them in sync with the releases until official packages are available.

**Comment by ftrojahn** [comment]
Thank you very much @nbuchwitz - site is working using http; https asks for htaccess login, thou ...

**Comment by nbuchwitz** [comment]
@ftrojahn fixed (little IPv6 issue)

**Comment by Codelica** [comment]
@nbuchwitz thank you!  Any chance you could be bribed for Ubuntu pipelines? Just curious. :)

**Comment by matt407** [comment]
Any chance to provide packages for arm64 / Debian, too?

**Comment by lazyfrosch** [comment]
We plan to build armhf packages for Raspbian in the future, but there is no date set.

If you have professional interest in packages and support of a certain platform, please contact us at info@icinga.com

**Comment by dasJ** [comment]
So I have finally managed to cross-compile Icinga 2 in Jenkins. Our installation tries to fetch the latest Icinga 2 release for amd64 and if a new version was released since the last build, a new package is built for arm and arm64.

To use the script, you'll need to have `binfmt_misc` loaded into your kernel. We execute the build itself inside a privileged Docker container (privileged to use binfmt) and execute the following script:

```
set +x
buildFor=(stretch:arm64 jessie:armhf)
binfmts=(arm aarch64)
mirror=http://deb.debian.org/debian

# Get rid of the man pages - speeds up apt installations
set +e
echo '
path-exclude=/usr/share/man/*
path-exclude=/usr/share/doc/*' > /etc/dpkg/dpkg.cfg.d/01_nodoc
rm -rf /usr/share/doc
rm -rf /usr/share/man
set -e

# Prepare apt (Icinga 2 repo and required packages)
osid="$(awk -F"=" '/^ID=/ {print $2}' /etc/os-release)"
osver="$(awk -F"[()]+" '/^VERSION=/ {print $2}' /etc/os-release)"
echo "deb-src http://packages.icinga.com/${osid} icinga-${osver} main" > /etc/apt/sources.list.d/icinga.list
wget -q -O - https://packages.icinga.com/icinga.key | apt-key add -
apt update
apt install -y ubuntu-dev-tools pbuilder qemu-user-static

# Prepare pbuilder and binfmt
echo 'PBUILDERSATISFYDEPENDSCMD="/usr/lib/pbuilder/pbuilder-satisfydepends-experimental"' >> /etc/pbuilderrc
for binfmt in "${binfmts[@]}"; do
	update-binfmts --enable "qemu-${binfmt}"
done

# Prepare sources
mkdir -p build out tmp
cd build
apt source icinga2

# Build targets
for target in "${buildFor[@]}"; do
	distro="${target%:*}"
    arch="${target#*:}"
    
    rm -rf ../tmp/*
    [ -f "/root/pbuilder/${distro}-${arch}-base.tgz" ] || pbuilder-dist "${distro}" "${arch}" create --mirror "${mirror}"
    DEB_BUILD_OPTIONS=nocheck pbuilder-dist "${distro}" "${arch}" *.dsc --buildresult "$(pwd)/../tmp"
    cp -v ../tmp/*.deb ../out
done
```

The code isn't really optimal because it uses qemu-user, but it comletes the arm64 build in ~1.5h. I hope this helps anyone (especially the Netways team).

**Comment by syswombat** [comment]
reading to stay informed

**Comment by MarcusCaepio** [comment]
Any news here? Just bought a Raspi to use Icinga2 for my home network and sitting here realising, that there are noh packages anymore. Trying to build it with docker....very annoying....

**Comment by snooops** [comment]
Hi @dasJ 

i tried your script on a fresh pi64bit Debian, which is a Debian Stretch. And im getting this error:
```
update-binfmts: warning: qemu-arm not in database of installed binary formats.
update-binfmts: exiting due to previous errors
```

**Comment by ftrojahn** [comment]
May be this helps?

https://wiki.debian.org/QemuUserEmulation

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by snooops** [comment]
Hmm, cant get it running. @dasJ could you provide your deb package? :)

**Comment by dasJ** [comment]
Ah yes, while these are only builds for armhf, I only have aarch64 builds. So the repository @jalogisch mentioned is probably the correct one for you

**Comment by nbuchwitz** [comment]
I have also current aarch64 builds for stretch if it helps.

**Comment by dnsmichi** [comment]
Builds for Debian Stretch fail mainly because of the weird gcc 6.3 compiler behaviour with our C++11 code base. There are changes in 2.9.x using more advanced coding techniques with std::map, initializer lists and so on.

During our research on the OSMC hackathon yesterday, we've found that the timer tests can be fixed with icinga/icinga2#6760 while the problem with dictionaries and object packer do create more memory access violations with gcc 6.3 on armhf. I do believe that this is a regression/bug in libstdc++6 or gcc itself on armhf, as we've read that raspbian keep using gcc 4.x for better compatibility.

This is tracked by icinga/icinga2#6761 and blocks any further package build attempts.

**Comment by nbuchwitz** [comment]
@jalogisch & others: With today's 2.10.2 release the builds for stretch work again. Would be great if you could test if these work for you. I'm especially interested in long term tests, as @dnsmichi and I ran into some problems with the GCC  6.3 compiler optimizations in our OSMC hackathon sesission.

**Comment by matt407** [comment]
Build for Version 2.10.2 unfortunately fails on an arm64 device (OS: Armbian Stretch (based on Debian))
Could you comprehend that - the build process worked fine with version 2.9.x and 2.10.1?


`[ 88%] Building CXX object test/CMakeFiles/boosttest-test-icinga_checkable.dir/icinga_checkable_unity.cpp.o
cd /build/icinga2-2.10.2/obj-aarch64-linux-gnu/test && /usr/bin/c++    -I/build/icinga2-2.10.2 -I/build/icinga2-2.10.2/lib -I/build/icinga2-2.10.2/obj-aarch64-linux-gnu -I/build/icinga2-2.10.2/obj-aarch64-linux-gnu/lib -I/build/icinga2-2.10.2/obj-aarch64-linux-gnu/test  -g -O2 -fdebug-prefix-map=/build/icinga2-2.10.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -pthread -Winvalid-pch -flto -std=c++0x   -o CMakeFiles/boosttest-test-icinga_checkable.dir/icinga_checkable_unity.cpp.o -c /build/icinga2-2.10.2/obj-aarch64-linux-gnu/test/icinga_checkable_unity.cpp
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-6/README.Bugs> for instructions.
lib/db_ido/CMakeFiles/db_ido.dir/build.make:125: recipe for target 'lib/db_ido/CMakeFiles/db_ido.dir/db_ido_unity.cpp.o' failed
make[4]: *** [lib/db_ido/CMakeFiles/db_ido.dir/db_ido_unity.cpp.o] Error 4
make[4]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
CMakeFiles/Makefile2:1897: recipe for target 'lib/db_ido/CMakeFiles/db_ido.dir/all' failed
make[3]: *** [lib/db_ido/CMakeFiles/db_ido.dir/all] Error 2
make[3]: *** Waiting for unfinished jobs....
[ 88%] Linking CXX executable ../Bin/None/boosttest-test-icinga_checkable
cd /build/icinga2-2.10.2/obj-aarch64-linux-gnu/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/boosttest-test-icinga_checkable.dir/link.txt --verbose=1
/usr/bin/c++   -g -O2 -fdebug-prefix-map=/build/icinga2-2.10.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -pthread -Winvalid-pch -flto -std=c++0x  -Wl,-z,relro -flto CMakeFiles/boosttest-test-icinga_checkable.dir/test-runner.cpp.o CMakeFiles/boosttest-test-icinga_checkable.dir/icinga_checkable_unity.cpp.o ../third-party/mmatch/CMakeFiles/mmatch.dir/mmatch.c.o ../third-party/socketpair/CMakeFiles/socketpair.dir/socketpair.c.o ../lib/base/CMakeFiles/base.dir/application-version.cpp.o ../lib/base/CMakeFiles/base.dir/base_unity.cpp.o ../third-party/execvpe/CMakeFiles/execvpe.dir/execvpe.c.o ../lib/config/CMakeFiles/config.dir/config_lexer.cc.o ../lib/config/CMakeFiles/config.dir/config_parser.cc.o ../lib/config/CMakeFiles/config.dir/config_unity.cpp.o ../lib/remote/CMakeFiles/remote.dir/remote_unity.cpp.o ../lib/icinga/CMakeFiles/icinga.dir/icinga-itl.cpp.o ../lib/icinga/CMakeFiles/icinga.dir/icinga_unity.cpp.o ../lib/cli/CMakeFiles/cli.dir/cli_unity.cpp.o  -o ../Bin/None/boosttest-test-icinga_checkable -rdynamic -ldl -lboost_thread -lboost_system -lboost_program_options -lboost_regex -lboost_chrono -lboost_date_time -lboost_atomic -lpthread -lssl -lcrypto -lyajl -lsystemd -ledit -ltermcap -lboost_unit_test_framework 
make[4]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
[ 88%] Built target perfdata
make[4]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
[ 88%] Built target check_nscp_api
make[4]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
`
`[ 88%] Built target boosttest-test-icinga_checkable
make[3]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
Makefile:163: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/build/icinga2-2.10.2/obj-aarch64-linux-gnu'
dh_auto_build: make -j4 returned exit code 2
debian/rules:33: recipe for target 'override_dh_auto_build' failed
make[1]: *** [override_dh_auto_build] Error 2
make[1]: Leaving directory '/build/icinga2-2.10.2'
debian/rules:28: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2`

**Comment by dnsmichi** [comment]
Typical OOM when compiling/linking unity builds on small sized platforms.

```
/build/icinga2-2.10.2/obj-aarch64-linux-gnu/test/icinga_checkable_unity.cpp c++: internal compiler error: Killed (program cc1plus) Please submit a full bug report, with preprocessed source if appropriate. See <file:///usr/share/doc/gcc-6/README.Bugs> for instructions.
```

Try to disable unity builds for embedded hardware, docker containers, Windows VMs.

**Comment by Jonny007-MKD** [comment]
Hi, did I understand correctly that armhf builds are back for Debian Stretch? Do you also have plans for Jessie?

**Comment by nbuchwitz** [comment]
For now you could use my builds of Jessie and stretch armhf packages (https://packages.enda.eu/).

**Comment by dnsmichi** [comment]
Here's the summary why you cannot build with gcc5/6 natively on armhf and without any unity build hacks: https://github.com/Icinga/icinga2/issues/6761#issuecomment-452757981

TL;DR: A regression with memory alignment in the compiler produced asm code for specific C++ basic types.

**Comment by lazyfrosch** [comment]
![good news everone](https://user-images.githubusercontent.com/121874/53191355-e8d00d80-360b-11e9-88d1-26b1f0031ce8.jpg)

Packages for Raspbian are available now, please see:

* https://icinga.com/2019/02/21/icinga-on-raspberry-pi/
* https://packages.icinga.com/raspbian/

**Comment by mzac** [comment]
Awesome!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by dnsmichi** [comment]
Cheers to ENDA KG & @nbuchwitz for sponsoring the work on this feature 🥇

**Comment by YtvwlD** [comment]
Is this specific to Raspbian or does it also work with eg. Ubuntu?

**Comment by dnsmichi** [comment]
That's specifically for Raspbian.
